apps/spark-3.5/src/test/java/com/linkedin/openhouse/catalog/e2e/SparkMoRFunctionalTest.java

```diff
@@ -0,0 +1,265 @@
+    sql("DROP TABLE IF EXISTS %s", tableName);
+  }
+
+  protected void initTable() {
+    sql(
+        "ALTER TABLE %s SET TBLPROPERTIES ('write.delete.mode'='merge-on-read', 'write.update.mode'='merge-on-read', 'write.merge.mode'='merge-on-read', 'write.delete.distribution-mode'='range');",
+        tableName);
+  }
+
+  protected void createAndInitTable(String schema) {
+    sql("CREATE TABLE openhouse.%s (%s) USING iceberg %s", tableName, schema, "");
+    initTable();
+  }
+
+  @Test
+  public void testDeleteFilesCanBeCreated() throws NoSuchTableException {
+    createAndInitTable("id int, data string");
+
+    List<SimpleRecord> records =
+        Arrays.asList(
+            new SimpleRecord(1, "a"),
+            new SimpleRecord(1, "b"),
+            new SimpleRecord(1, "c"),
+            new SimpleRecord(2, "d"),
+            new SimpleRecord(2, "e"));
+    ops.spark()
+        .createDataset(records, Encoders.bean(SimpleRecord.class))
+        .coalesce(1)
+        .writeTo(tableName)
+        .append();
+
+    sql("DELETE FROM %s WHERE id = 1 and data='a'", tableName);
+    sql("DELETE FROM %s WHERE id = 2 and data='d'", tableName);
+    sql("DELETE FROM %s WHERE id = 1 and data='c'", tableName);
+
+    IcebergTableStats stats = ops.collectTableStats(tableName);
+    assertThat(stats.getNumPositionDeleteFiles() + stats.getNumEqualityDeleteFiles()).isEqualTo(3L);
+
+    List<Object[]> expected = Arrays.asList(row(1, "b"), row(2, "e"));
+    List<Object[]> actual = sql("SELECT * FROM %s ORDER BY id ASC", tableName);
+    assertThat(actual).containsExactlyElementsOf(expected);
+  }
+
+  @Test
+  public void testCompactionCanRemovePositionDeleteFiles() throws NoSuchTableException {
+    createAndInitTable("id int, data string");
+
+    List<SimpleRecord> records =
+        Arrays.asList(
+            new SimpleRecord(1, "a"),
+            new SimpleRecord(1, "b"),
+            new SimpleRecord(1, "c"),
+            new SimpleRecord(2, "d"),
+            new SimpleRecord(2, "e"));
+    ops.spark()
+        .createDataset(records, Encoders.bean(SimpleRecord.class))
+        .coalesce(1)
+        .writeTo(tableName)
+        .append();
+
+    sql("DELETE FROM %s WHERE id = 1 and data='a'", tableName);
+    sql("DELETE FROM %s WHERE id = 2 and data='d'", tableName);
+    sql("DELETE FROM %s WHERE id = 1 and data='c'", tableName);
+
+    Table table = ops.getTable(tableName);
+
+    // first run of compaction
+    RewriteDataFiles.Result result = rewriteFunc.apply(ops, table);
+    Assertions.assertEquals(1, result.addedDataFilesCount());
+    Assertions.assertEquals(1, result.rewrittenDataFilesCount());
+
+    // this asserts the dangling delete problem, since compaction did NOT remove the final delete
+    IcebergTableStats stats = ops.collectTableStats(tableName);
+    assertThat(stats.getNumPositionDeleteFiles()).isEqualTo(3L);
+    assertThat(stats.getNumCurrentSnapshotPositionDeleteFiles()).isEqualTo(1L);
```

Comment:
for my understanding: why is there still a pos-delete file here as part of the current snapshot; also why is this pos-delete, which process generates this and how does that process determine between pos vs. eq?

Reply:
> why is there still a pos-delete file here as part of the current snapshot

delete here is dangling due to compaction generating a new datafile with the same data sequence number as the delete file. because the resulting datafile and delete file have same sequence number, the delete file isn't chosen to be dropped.

https://github.com/apache/iceberg/issues/4127


> why is this pos-delete

engine specific decision to create position delete. as far as i know, spark does not actually create equality deletes at all (only flink, and kafka connect)

Reply:
for the first question: but why does delete files become one although metadata-compaction is not called?

Reply:
compaction is called on line 131 above.

1. insert 3 deletes (shown in test)
2. table has 3 deletes associated with the current snapshot (this is not asserted, i will add it though for clarity)
3. compaction first time
4. only 1 delete associated with current snpashot: first 2 delete files were dropped due to being older than the newest datafile created by compaction in step 3, last 1 delete file is dangling